cloudinit/net/__init__.py

```diff
@@ -971,18 +972,38 @@ def get_ib_hwaddrs_by_interface():
     return ret
 
 
-def has_url_connectivity(url):
-    """Return true when the instance has access to the provided URL
+def has_url_connectivity(
+    url_data: Union[MutableMapping[str, Any], str]
+) -> bool:
+    """Return true when the instance has access to the provided URL.
 
     Logs a warning if url is not the expected format.
+
+    url_data is either the URL string to check, or a dictionary of kwargs
+    to send to readurl. E.g.:
+
+    has_url_connectivity("http://example.invalid")
+    has_url_connectivity({
+        "url": "http://example.invalid",
+        "headers": {"some": "header"},
+        "timeout": 10
+    })
     """
+    if isinstance(url_data, str):
+        url_data = {'url': url_data}
+    if 'url' not in url_data:
+        LOG.warning(
+            "Ignoring connectivity check. No 'url' to check in %s", url_data)
```

Comment:
Probably worth a return False here otherwise we will raise KeyError on the url_data['url'] attempt below.
```suggestion
        LOG.warning(
            "Ignoring connectivity check. No 'url' to check in %s", url_data)
        return False
```